raft/src/main/java/org/apache/kafka/raft/ReplicatedStateMachine.java

```diff
@@ -24,19 +24,22 @@
  * record appends will be routed to the leader, which will have an opportunity to either
  * accept or reject the append attempt.
  */
-public interface DistributedStateMachine {
+public interface ReplicatedStateMachine extends AutoCloseable {
 
     /**
      * Become a leader. This is invoked after a new election in the quorum if this
      * node was elected as the leader.
      *
      * @param epoch The latest quorum epoch
+     * @param recordAppender The record appender for appending records
      */
-    void becomeLeader(int epoch);
+    void becomeLeader(int epoch, RecordAppender recordAppender);
 
     /**
      * Become a follower. This is invoked after a new election finishes if this
-     * node was not elected as the leader.
+     * node was not elected as the leader. Note that only leader should have the access
```

Comment:
Thinking about the `becomeFollower` API a little bit. I think the current semantics are to wait until the election has finished. This means that the state machine may continue to write records while the election is going on. I wonder if it would be more useful if the two events are `onLeaderPromotion` and `onLeaderDemotion`. Then when the election starts, we can fire `onLeaderDemotion` immediately so that the state machine is not trying to write while the election is going on.

Reply:
That sounds like a better semantic than leader/follower transition.